bdns/servers.go

```diff
@@ -0,0 +1,159 @@
+// different strategies for providing addresses, and may provide different kinds
+// of addresses (e.g. host:port combos vs IP addresses).
+type ServerProvider interface {
+	Addrs() ([]string, error)
+	Stop()
+}
+
+// staticProvider stores a list of host:port combos, and provides that whole
+// list in randomized order when asked for addresses. This replicates the old
+// behavior of the bdns.impl's servers field.
+type staticProvider struct {
+	servers []string
+}
+
+var _ ServerProvider = &staticProvider{}
+
+func NewStaticProvider(servers []string) *staticProvider {
+	return &staticProvider{servers: servers}
+}
+
+func (sp *staticProvider) Addrs() ([]string, error) {
+	if sp.servers == nil || len(sp.servers) == 0 {
+		return nil, fmt.Errorf("no servers configured")
+	}
+	r := make([]string, len(sp.servers))
+	perm := rand.Perm(len(sp.servers))
+	for i, v := range perm {
+		r[i] = sp.servers[v]
+	}
+	return r, nil
+}
+
+func (sp *staticProvider) Stop() {}
+
+// dynamicProvider uses DNS to look up the set of IP addresses which correspond
+// to its single host. It returns this list in random order when asked for
+// addresses, and refreshes it regularly using a goroutine started by its
+// constructor.
+type dynamicProvider struct {
+	host string
+	// A map of IP addresses (Target fields in SRV records) to ports (Port fields
+	// in SRV records) associated with those addresses.
+	addrs  map[string][]uint16
+	mu     sync.RWMutex
+	cancel chan interface{}
+}
+
+var _ ServerProvider = &dynamicProvider{}
+
+// StartDynamicProvider constructs a new dynamicProvider and starts its
+// auto-update goroutine. The auto-update process queries DNS for SRV records
+// at refresh intervals and uses the resulting IP/port combos to populate the
+// list returned by Addrs. The update process ignores the Priority and Weight
+// attributes of the SRV records.
+func StartDynamicProvider(server string, refresh time.Duration) (*dynamicProvider, error) {
+	if server == "" {
+		return nil, fmt.Errorf("no DNS host provided")
+	}
+	updateCounter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "dns_update",
+			Help: "Counter of attempts to update a dynamic provider",
+		},
+		[]string{"success"},
+	)
+	dp := dynamicProvider{
+		host:   server,
+		addrs:  make(map[string][]uint16),
+		cancel: make(chan interface{}),
+	}
+	err := dp.update()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start dynamic provider: %w", err)
+	}
+
+	go func() {
+		t := time.NewTicker(refresh)
+		for {
+			select {
+			case <-t.C:
+				err := dp.update()
+				if err != nil {
+					updateCounter.With(prometheus.Labels{
+						"success": "false",
+					}).Inc()
+					continue
+				}
+				updateCounter.With(prometheus.Labels{
+					"success": "true",
+				}).Inc()
+			case <-dp.cancel:
+				return
+			}
+		}
+	}()
+
+	return &dp, nil
+}
+
+func (dp *dynamicProvider) update() error {
+	_, srvs, err := net.LookupSRV("dns", "", dp.host)
```

Comment:
The second argument here, `proto`, should be "udp" or "tcp". "udp" for our purposes.

Also let's use [DefaultResolver's LookupSRV method](https://golang.org/pkg/net/#Resolver.LookupSRV), which supports a context so we can time out.

Reply:
Oops thanks. The sd-test-srv doesn't care about the protocol, so I forgot that it would be important in production.